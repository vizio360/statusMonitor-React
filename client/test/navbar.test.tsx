import React from 'react';
import renderer from 'react-test-renderer';
import NavBar, {INavBarProps} from '@app/navbar';
import {render, fireEvent, waitForElement} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

describe('Nav Bar', () => {
  const renderNavBar = (
    props: INavBarProps = {dataChanged: false, editing: false},
  ) => {
    return render(
      <NavBar
        dataChanged={props.dataChanged}
        editing={props.editing}
        onEdit={props.onEdit}
        onCancel={props.onCancel}
        onSave={props.onSave}
        onAddNode={props.onAddNode}
      />,
    );
  };

  it('initiliases correctly with the edit button only visible', () => {
    let navBar = renderer.create(
      <NavBar dataChanged={false} editing={false} />,
    );
    let navBarJson = navBar.toJSON();
    expect(navBarJson).toMatchSnapshot();
  });

  it('renders the editor menu when starting editing', () => {
    const {getByText, asFragment} = renderNavBar({
      dataChanged: false,
      editing: true,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the edit button if not editing', () => {
    const {getByText, asFragment} = renderNavBar();
    expect(asFragment()).toMatchSnapshot();
  });

  it('disables the Save button if in edit mode and no changes have been made', () => {
    const {getByText, asFragment} = renderNavBar({
      dataChanged: false,
      editing: true,
    });
    expect(getByText('Save')).toBeDisabled();
  });

  it('enables the Save button is in edit mode and changes have been made', () => {
    const {getByText, asFragment} = renderNavBar({
      dataChanged: true,
      editing: true,
    });
    expect(getByText('Save')).toBeEnabled();
  });

  it('notifies the onEdit callback when editing starts', () => {
    const onEditCB = jest.fn();
    const {getByText, asFragment} = renderNavBar({
      dataChanged: false,
      editing: false,
      onEdit: onEditCB,
    });
    fireEvent.click(getByText('Edit'));
    expect(onEditCB).toHaveBeenCalled();
  });

  it('notifies the onCancel callback when editing is cancelled', () => {
    const onCancelCB = jest.fn();
    const {getByText, asFragment} = renderNavBar({
      dataChanged: false,
      editing: true,
      onCancel: onCancelCB,
    });
    fireEvent.click(getByText('Cancel'));
    expect(onCancelCB).toHaveBeenCalled();
  });

  it('notifies the onSave callback when editing is saved', () => {
    const onSaveCB = jest.fn();
    const {getByText, asFragment} = renderNavBar({
      dataChanged: true,
      editing: true,
      onSave: onSaveCB,
    });
    fireEvent.click(getByText('Save'));
    expect(onSaveCB).toHaveBeenCalled();
  });

  it('notifies the onAddNode callback when editing and adding a node', () => {
    const onAddNodeCB = jest.fn();
    const {getByText, asFragment} = renderNavBar({
      dataChanged: false,
      editing: true,
      onAddNode: onAddNodeCB,
    });
    fireEvent.click(getByText('Add Node'));
    expect(onAddNodeCB).toHaveBeenCalled();
  });
});
