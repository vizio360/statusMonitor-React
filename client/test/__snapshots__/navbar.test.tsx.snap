// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Bar initiliases correctly with the edit button only visible 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-light bg-light"
>
  <a
    className="navbar-brand"
    href="#"
  >
    Monitor
  </a>
  <button
    aria-controls="navbarSupportedContent"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#navbarSupportedContent"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse"
    id="navbarSupportedContent"
  >
    <ul
      className="navbar-nav mr-auto"
    >
      <li
        className="nav-item active"
      >
        <button
          className="btn btn-primary mr-sm-2"
          onClick={[Function]}
          type="button"
        >
          Edit
        </button>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`Nav Bar renders the edit button if not editing 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="#"
    >
      Monitor
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      class="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        class="navbar-nav mr-auto"
      >
        <li
          class="nav-item active"
        >
          <button
            class="btn btn-primary mr-sm-2"
            type="button"
          >
            Edit
          </button>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Nav Bar renders the editor menu when starting editing 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="#"
    >
      Monitor
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      class="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        class="navbar-nav mr-auto"
      >
        <li
          class="nav-item active"
        >
          <button
            class="btn btn-primary mr-sm-2"
            type="button"
          >
            Add Node
          </button>
        </li>
        <li
          class="nav-item"
        >
          <button
            class="btn btn-primary mr-sm-2"
            disabled=""
            type="button"
          >
            Save
          </button>
        </li>
        <li
          class="nav-item"
        >
          <button
            class="btn btn-primary mr-sm-2"
            type="button"
          >
            Cancel
          </button>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;
